In [314]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Задание 1

Исследуйте систему векторов на линейную зависимость или независимость без использования Python (вручную): 

$\overrightarrow{a_1} = (1; 8; -1), \overrightarrow{a_2} = (-2; 3; 3), \overrightarrow{a_3} = (4; -11; 9)$

Подробно распишите решение, а в итоге сделайте вывод про то, являются ли векторы зависимыми или независимыми. 

# Решение

Векторы $\overrightarrow{a_1}, \overrightarrow{a_2}, \overrightarrow{a_3}$ линейно зависимы, если $\exists$ $\omega_1, \omega_2, \omega_3$ и не все они равны нулю, что: $\omega_1 \overrightarrow{a_1} + \omega_2 \overrightarrow{a_2} + \omega_3 \overrightarrow{a_3} = \overrightarrow{0}$

$
A =
\begin{pmatrix}
   1 & -2 & 4 \\
   8 & 3 & -11 \\
   -1 & 3 & 9 \\
\end{pmatrix}$, 
$
\omega =
\begin{pmatrix}
   \omega_1 \\
   \omega_2 \\
   \omega_3 \\
\end{pmatrix}
$, тогда система в матричном виде:

$A\overrightarrow{w}=\overrightarrow{0}$

$\begin{pmatrix}
   1 & -2 & 4 \\
   8 & 3 & -11 \\
   -1 & 3 & 9 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
   \omega_1 \\
   \omega_2 \\
   \omega_3 \\
\end{pmatrix}
=\begin{pmatrix}
   0 \\
   0 \\
   0 \\
\end{pmatrix}$

Перед нами однородная СЛУ. У неё уже есть тривиальное решение $\omega_1=\omega_2=\omega_3=0$. Узнаем ранг матрицы с помощью метода Гаусса. Матрица системы:

$
\begin{pmatrix}
   1 & -2 & 4 \\
   8 & 3 & -11 \\
   -1 & 3 & 9 \\
\end{pmatrix}
\begin{matrix}
(1) \\
(2) \\
(3) \\
\end{matrix}
$

Умножаем (1) на -8 и прибавляем ко (2):

$
\begin{pmatrix}
   1 & -2 & 4 \\
   0 & 19 & -43 \\
   -1 & 3 & 9 \\
\end{pmatrix}
\begin{matrix}
(1) \\
(2) \\
(3) \\
\end{matrix}
$

Прибавляем (1) к (3):

$
\begin{pmatrix}
   1 & -2 & 4 \\
   0 & 19 & -43 \\
   0 & 1 & 13 \\
\end{pmatrix}
\begin{matrix}
(1) \\
(2) \\
(3) \\
\end{matrix}
$

Умножаем (3) на -19 и прибавляем к нему (2):

$
\begin{pmatrix}
   1 & -2 & 4 \\
   0 & 19 & -43 \\
   0 & 0 & -290 \\
\end{pmatrix}
\begin{matrix}
(1) \\
(2) \\
(3) \\
\end{matrix}
$

Матрица приведена к ступенчатому виду. Ранг матрицы равен 3 - количеству переменных. Значит эта однородная СЛУ имеет только тривиальное решение. 

Следовательно векторы $\overrightarrow{a_1}, \overrightarrow{a_2}, \overrightarrow{a_3}$ линейно независимы.

Ответ: векторы $\overrightarrow{a_1}, \overrightarrow{a_2}, \overrightarrow{a_3}$ линейно независимы.

# Задание 2

При каких значениях параметра $a$ система векторов является линейно зависимой? Найдите решение без использования Python (вручную). 

$\overrightarrow{a_1} = (1; 0; 2; -3), \overrightarrow{a_2} = (2; 2; -4; a), \overrightarrow{a_3} = (3; 1; 2; -5)$

# Решение

Векторы $\overrightarrow{a_1}, \overrightarrow{a_2}, \overrightarrow{a_3}$ линейно зависимы, если $\exists$ $\omega_1, \omega_2, \omega_3$ и не все они равны нулю, что: $\omega_1 \overrightarrow{a_1} + \omega_2 \overrightarrow{a_2} + \omega_3 \overrightarrow{a_3} = \overrightarrow{0}$

$
A =
\begin{pmatrix}
   1 & 2 & 3 \\
   0 & 2 & 1 \\
   2 & -4 & 2 \\
   -3 & a &-5 \\
\end{pmatrix}$, 
$
\omega =
\begin{pmatrix}
   \omega_1 \\
   \omega_2 \\
   \omega_3 \\
\end{pmatrix}
$, тогда система в матричном виде:

$A\overrightarrow{w}=\overrightarrow{0}$

$\begin{pmatrix}
   1 & 2 & 3 \\
   0 & 2 & 1 \\
   2 & -4 & 2 \\
   -3 & a &-5 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
   \omega_1 \\
   \omega_2 \\
   \omega_3 \\
\end{pmatrix}
=\begin{pmatrix}
   0 \\
   0 \\
   0 \\
\end{pmatrix}$

Если $rank A = r$, то существует хотя бы одна $r \times r$ подматрица с ненулевым определителем. Для того, чтобы система векторов была линейно зависимой ранг матрицы системы должен быть меньше 3, а следовательно все $3 \times 3$ подматрицы должны иметь нулевой определитель. 

Первая подматрица:

$det
\begin{pmatrix}
   1 & 2 & 3 \\
   0 & 2 & 1 \\
   2 & -4 & 2 \\
\end{pmatrix}=
1\cdot
\begin{vmatrix}
   2 & 1 \\
   -4 & 2
\end{vmatrix}
-2\cdot
\begin{vmatrix}
   0 & 1 \\
   2 & 2
\end{vmatrix}
+3\cdot
\begin{vmatrix}
   0 & 2 \\
   2 & -4
\end{vmatrix}
= 8 + 4 - 12 = 0
$. Условие выполнено.

Вторая подматрица:

$det
\begin{pmatrix}
   0 & 2 & 1 \\
   2 & -4 & 2 \\
   -3 & a &-5 \\
\end{pmatrix}=
1\cdot
\begin{vmatrix}
   2 & 1 \\
   a & -5
\end{vmatrix}
-2\cdot
\begin{vmatrix}
   0 & 1 \\
   -3 & -5
\end{vmatrix}
+3\cdot
\begin{vmatrix}
   0 & 2 \\
   -3 & a
\end{vmatrix}
= -10 - a + 6 + 18 = 2 - a
$

Чтобы определитель второй подматрицы был нулем $a = 2$.

Ответ: при $a = 2$ система векторов является линейно зависимой.

# Задание 3

Выразите определитель матрицы. Найдите решение без использования Python (вручную). 

$
\begin{vmatrix}
    1&-1&a&-1 \\
    -1&-2&b&-1 \\
    -2&0&c&1 \\
    0&1&d&0 \\
\end{vmatrix}
$

# Решение

По первой строке:

$
\det = 1 \cdot 
\begin{vmatrix}
    -2 & b & -1 \\
     0 & c &  1 \\
     1 & d &  0 \\
\end{vmatrix}
-(-1) \cdot
\begin{vmatrix}
    -1 & b & -1 \\
    -2 & c &  1 \\
     0 & d &  0 \\
\end{vmatrix}
+a \cdot
\begin{vmatrix}
    -1 & -2 & -1 \\
    -2 &  0 &  1 \\
     0 &  1 &  0 \\
\end{vmatrix}
-(-1) \cdot
\begin{vmatrix}
    -1 & -2 & b \\
    -2 &  0 & c \\
     0 &  1 & d \\
\end{vmatrix}
$

---

Первый минор:

$
\begin{vmatrix}
    -2 & b & -1 \\
     0 & c &  1 \\
     1 & d &  0 \\
\end{vmatrix}
= -2 \cdot
\begin{vmatrix}
    c &  1 \\
    d &  0 \\
\end{vmatrix}
-b \cdot
\begin{vmatrix}
     0 &  1 \\
     1 &  0 \\
\end{vmatrix}
+(-1) \cdot
\begin{vmatrix}
     0 & c \\
     1 & d \\
\end{vmatrix}
= 2d + b + c
$

---

Второй минор:

$
\begin{vmatrix}
    -1 & b & -1 \\
    -2 & c &  1 \\
     0 & d &  0 \\
\end{vmatrix}
= -1 \cdot \begin{vmatrix}
    c &  1 \\
    d &  0 \\
\end{vmatrix}
-b \cdot \begin{vmatrix}
    -2 &  1 \\
     0 &  0 \\
\end{vmatrix}
+(-1) \cdot \begin{vmatrix}
    -2 & c \\
     0 & d \\
\end{vmatrix}
= d + 2d
= 3d
$

---

Третий минор:

$
\begin{vmatrix}
    -1 & -2 & -1 \\
    -2 &  0 &  1 \\
     0 &  1 &  0 \\
\end{vmatrix}
= -1 \cdot \begin{vmatrix}
    0 &  1 \\
    1 &  0 \\
\end{vmatrix}
-(-2) \cdot \begin{vmatrix}
    -2 &  1 \\
     0 &  0 \\
\end{vmatrix}
+(-1) \cdot \begin{vmatrix}
    -2 &  0 \\
     0 &  1 \\
\end{vmatrix}
= 1 + 0 + 2
= 3
$


---

Четвертый минор:

$
\begin{vmatrix}
    -1 & -2 & b \\
    -2 &  0 & c \\
     0 &  1 & d \\
\end{vmatrix}
= -1 \cdot \begin{vmatrix}
    0 & c \\
    1 & d \\
\end{vmatrix}
-(-2) \cdot \begin{vmatrix}
    -2 & c \\
     0 & d \\
\end{vmatrix}
+b \cdot \begin{vmatrix}
    -2 & 0 \\
     0 & 1 \\
\end{vmatrix}
= c - 4d - 2b
$

---

Подставляем:

$
\det = 1 \cdot (2d + b + c) - (-1) \cdot 3d + a \cdot 3 - (-1) \cdot (c - 4d - 2b) = d - b + 2c + 3a
$

Ответ: $
\begin{vmatrix}
    1&-1&a&-1 \\
    -1&-2&b&-1 \\
    -2&0&c&1 \\
    0&1&d&0 \\
\end{vmatrix}
=d - b + 2c + 3a
$

# Задание 4

Найдите значения параметров a, b и c, при которых матрицы A и B являются обратными. Найдите решение без использования Python (вручную). 

$
A =
\begin{pmatrix}
   a & -2 & -1 \\
   -15 & b+20 & 12 \\
   10 & -19 & 2c \\
\end{pmatrix}, 
B =
\begin{pmatrix}
   -4 & 3 & 5 \\
   0 & 2 & 3 \\
   -5 & -1 & -1 \\
\end{pmatrix}
$

# Решение

$A$ и $B$ являются обратными, если $AB=E = 
\begin{pmatrix}
   1 & 0 & 0 \\
   0 & 1 & 0 \\
   0 & 0 & 1 \\
\end{pmatrix}
$

$AB=
\begin{pmatrix}
   a & -2 & -1 \\
   -15 & b+20 & 12 \\
   10 & -19 & 2c \\
\end{pmatrix}
\cdot
\begin{pmatrix}
   -4 & 3 & 5 \\
   0 & 2 & 3 \\
   -5 & -1 & -1 \\
\end{pmatrix} =
\begin{pmatrix}
   -4a + 5 & 3a - 3 & 5a - 5 \\
   0 & 2b - 17 & 3b - 27 \\
   -40 - 10c & -8 - 2c & -7 - 2c \\
\end{pmatrix} =
\begin{pmatrix}
   1 & 0 & 0 \\
   0 & 1 & 0 \\
   0 & 0 & 1 \\
\end{pmatrix}
$

Находим $a, b, c$:

$-4a + 5 = 1 \implies a = 1$

$2b - 17 = 1 \implies b = 9$

$-7-2c = 1 \implies c = -4$

Ответ: $a = 1, b = 9, c = -4$

# Задание 5

Найдите собственные векторы и собственные значения матрицы. Найдите решение без использования Python (вручную). 

$
\begin{pmatrix}
   4 & 0 & -1 \\
   2 & 2 & 1 \\
   -1 & 0 & 4 \\
\end{pmatrix}
$

# Решение

Обозначим матрицу за $A$.

### 1. Собственные значения

Характеристическое уравнение:

$det(A - \lambda E) = 0$

$det(A - \lambda E) =
\begin{vmatrix}
   4 - \lambda & 0 & -1 \\
   2 & 2 - \lambda & 1 \\
   -1 & 0 & 4 - \lambda \\
\end{vmatrix} = - \lambda^3 +10\lambda^2-31\lambda+30 = 
-(\lambda-2)(\lambda^2-8\lambda+15)=-(\lambda-2)(\lambda-3)(\lambda-5)=0$

Следовательно собственные числа:

$\lambda_1 = 2$

$\lambda_2 = 3$

$\lambda_3 = 5$

### 2. Собственные векторы

Для $\lambda = 2$:

$(A - 2E)v_1 =
\begin{pmatrix}
   2 & 0 & -1 \\
   2 & 0 & 1 \\
   -1 & 0 & 2 \\
\end{pmatrix}v_1
= 0$

Система уравнений:

$
\begin{cases}
   2x - z = 0 \\
   2x + z = 0 \\
   -x + 2z = 0
\end{cases}
$

$\implies x = 0, z = 0$. $y$ может быть любым, возьмём 1. $v_1=
\begin{pmatrix}
   0 \\
   1 \\
   0 \\
\end{pmatrix}
$

---

Для $\lambda = 3$:

$(A - 3E)v_2 =
\begin{pmatrix}
   1 & 0 & -1 \\
   2 & -1 & 1 \\
   -1 & 0 & 1 \\
\end{pmatrix}v_2
= 0$

Система уравнений:

$
\begin{cases}
   x - z = 0 \\
   2x - y + z = 0 \\
   -x + z = 0 \\
\end{cases}
$

$\implies x = z, y = 3z$. Возьмём $z$ = 1. $v_2=
\begin{pmatrix}
   1 \\
   3 \\
   1 \\
\end{pmatrix}
$

---

Для $\lambda = 5$:

$(A - 5E)v_3 =
\begin{pmatrix}
   -1 & 0 & -1 \\
   2 & -3 & 1 \\
   -1 & 0 & -1 \\
\end{pmatrix}v_3
= 0$

Система уравнений:

$
\begin{cases}
   -x - z = 0 \\
   2x - 3y + z = 0 \\
   -x - z = 0
\end{cases}
$

$\implies x = -z, y = -\dfrac{1}{3}z$. Возьмём $z$ = 3. $v_3=
\begin{pmatrix}
   -3 \\
   -1 \\
   3 \\
\end{pmatrix}
$

---

### Собственные векторы и значения

1. $\lambda = 2, v_1 = 
\begin{pmatrix}
   0 \\
   1 \\
   0 \\
\end{pmatrix}
$
2. $\lambda = 3, v_2 = 
\begin{pmatrix}
   1 \\
   3 \\
   1 \\
\end{pmatrix}
$
3. $\lambda = 5, v_3 = 
\begin{pmatrix}
   -3 \\
   -1 \\
   3 \\
\end{pmatrix}
$

# Задание 6

Найдите сумму сингулярных чисел для матрицы: 

$
\begin{pmatrix}
   -2 & 0 & 1 \\
   0 & -2 & 1 \\
   1 & 1 & -1 \\
\end{pmatrix}
$

Найдите решение без использования Python (вручную). 

# Решение

Найдем $A^TA$:

$ A^TA=
\begin{pmatrix}
   -2 & 0 & 1 \\
   0 & -2 & 1 \\
   1 & 1 & -1 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
   -2 & 0 & 1 \\
   0 & -2 & 1 \\
   1 & 1 & -1 \\
\end{pmatrix} =
\begin{pmatrix}
   5 & 1 & -3 \\
   1 & 5 & -3 \\
   -3 & -3 & 3 \\
\end{pmatrix}
$

---

Найдем собственные числа. Характеристическое уравнение:

$det(A^TA - \lambda E) = 0$:

$det(A^TA - \lambda E) =
\begin{vmatrix}
   5 - \lambda & 1 & -3 \\
   1 & 5 - \lambda & -3 \\
   -3 & -3 & 3 - \lambda \\
\end{vmatrix}=
-\lambda^3 + 13\lambda^2-36\lambda=-\lambda(\lambda^2-13\lambda+36) = -\lambda(\lambda-4)(\lambda-9)=0
$

$
\sigma_1 = \sqrt{0} = 0,
$

$
\sigma_2 = \sqrt{4} = 2,
$

$
\sigma_3 = \sqrt{9} = 3.
$

$
\sigma_1 + \sigma_2 + \sigma_3 = 0 + 2 + 3 = 5.
$

Ответ: сумма сингулярных чисел для матрицы равна 5.

# Часть 2: Задания, решаемые с помощью Python

Даны данные об аренде велосипедов. 

Описание переменных:

* season: 1 - весна, 2 - лето, 3 - осень, 4 - зима
* yr: 0 - 2011, 1 - 2012
* mnth: месяц, от 1 до 12
* holiday: 0 - нет праздника, 1 - есть праздник
* weekday: день недели, от 0 до 6
* workingday: 0 - нерабочий день, 1 - рабочий день
* weathersit: оценка благоприятности погоды от 1 (чистый, ясный день) до 4 (ливень, туман)
* temp: температура в Цельсиях
* atemp: температура по ощущениям в Цельсиях
* hum: влажность
* windspeed(mph): скорость ветра в милях в час
* windspeed(ms): скорость ветра в метрах в секунду
* cnt: количество арендованных велосипедов (это целевой признак, его мы будем предсказывать)


# Задание 7

Загрузите данные. Изучите, есть ли в них пропущенные значения? Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода. 

In [323]:
df = pd.read_csv('bikes_rent.xls')
df

season  yr  mnth  holiday  weekday  workingday  weathersit       temp  \
0         1   0     1        0        6           0           2  14.110847   
1         1   0     1        0        0           0           2  14.902598   
2         1   0     1        0        1           1           1   8.050924   
3         1   0     1        0        2           1           1   8.200000   
4         1   0     1        0        3           1           1   9.305237   
..      ...  ..   ...      ...      ...         ...         ...        ...   
726       1   1    12        0        4           1           2  10.420847   
727       1   1    12        0        5           1           2  10.386653   
728       1   1    12        0        6           0           2  10.386653   
729       1   1    12        0        0           0           1  10.489153   
730       1   1    12        0        1           1           2   8.849153   

        atemp      hum  windspeed(mph)  windspeed(ms)   cnt  
0    18.18125  80.5833       10.749882       4.805490   985  
1    17.68695  69.6087       16.652113       7.443949   801  
2     9.47025  43.7273       16.636703       7.437060  1349  
3    10.60610  59.0435       10.739832       4.800998  1562  
4    11.46350  43.6957       12.522300       5.597810  1600  
..        ...      ...             ...            ...   ...  
726  11.33210  65.2917       23.458911      10.486773  2114  
727  12.75230  59.0000       10.416557       4.656485  3095  
728  12.12000  75.2917        8.333661       3.725374  1341  
729  11.58500  48.3333       23.500518      10.505372  1796  
730  11.17435  57.7500       10.374682       4.637766  2729  

[731 rows x 13 columns]

In [324]:
df.isnull().sum()

season            0
yr                0
mnth              0
holiday           0
weekday           0
workingday        0
weathersit        0
temp              0
atemp             0
hum               0
windspeed(mph)    0
windspeed(ms)     0
cnt               0
dtype: int64

В данных нет пропущенных значений.

# Задание 8

Исследуем данные на наличие линейной зависимости между столбцами. Найдите ранг матрицы данных. Сравните его с количеством столбцов, сделайте вывод. Если линейная зависимость есть - напишите, между какими столбцами и удалите один из них.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода. 

In [327]:
matrix_rank = np.linalg.matrix_rank(df)
columns_count = df.shape[1]
print(f'Ранг матрицы: {matrix_rank}')
print(f'Количество столбцов: {columns_count}')

Ранг матрицы: 12
Количество столбцов: 13


Есть линейная зависимость

In [329]:
# нахождение матрицы корреляции и удаление недостаточно высоких значений
corr_m = df.corr()
high_corr = corr_m[(corr_m > 0.8) & (corr_m < 1.0)]
high_corr = high_corr.dropna(axis=0, how="all")
high_corr = high_corr.dropna(axis=1, how="all")
high_corr

season     mnth      temp     atemp  windspeed(mph)  \
season              NaN  0.83144       NaN       NaN             NaN   
mnth            0.83144      NaN       NaN       NaN             NaN   
temp                NaN      NaN       NaN  0.991702             NaN   
atemp               NaN      NaN  0.991702       NaN             NaN   
windspeed(mph)      NaN      NaN       NaN       NaN             NaN   
windspeed(ms)       NaN      NaN       NaN       NaN             1.0   

                windspeed(ms)  
season                    NaN  
mnth                      NaN  
temp                      NaN  
atemp                     NaN  
windspeed(mph)            1.0  
windspeed(ms)             NaN

Из матрицы видно, что значения корреляции между temp и atemp а также windspeed(mph) и windspeed(ms) довольно высока. У скорости ветра прямая зависимость.

Удалям столбец windspeed(ms)

In [331]:
df = df.drop(columns='windspeed(ms)')

# Задание 9

Выведите корреляционную матрицу оставшихся переменных.

* Какая переменная сильнее всего коррелирует с целевой?
* Есть ли мультиколлинеарность? Между какими переменными? Если есть корреляция выше 0.9 - оставьте только 1 признак. Объясните выбор

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода. 

In [333]:
corr_m = df.corr()
corr_m

season        yr      mnth   holiday   weekday  workingday  \
season          1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr             -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth            0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday        -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday        -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday      0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit      0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp            0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp           0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum             0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(mph) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt             0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

                weathersit      temp     atemp       hum  windspeed(mph)  \
season            0.019211  0.334315  0.342876  0.205445       -0.229046   
yr               -0.048727  0.047604  0.046106 -0.110651       -0.011817   
mnth              0.043528  0.220205  0.227459  0.222204       -0.207502   
holiday          -0.034627 -0.028556 -0.032507 -0.015937        0.006292   
weekday           0.031087 -0.000170 -0.007537 -0.052232        0.014282   
workingday        0.061200  0.052660  0.052182  0.024327       -0.018796   
weathersit        1.000000 -0.120602 -0.121583  0.591045        0.039511   
temp             -0.120602  1.000000  0.991702  0.126963       -0.157944   
atemp            -0.121583  0.991702  1.000000  0.139988       -0.183643   
hum               0.591045  0.126963  0.139988  1.000000       -0.248489   
windspeed(mph)    0.039511 -0.157944 -0.183643 -0.248489        1.000000   
cnt              -0.297391  0.627494  0.631066 -0.100659       -0.234545   

                     cnt  
season          0.406100  
yr              0.566710  
mnth            0.279977  
holiday        -0.068348  
weekday         0.067443  
workingday      0.061156  
weathersit     -0.297391  
temp            0.627494  
atemp           0.631066  
hum            -0.100659  
windspeed(mph) -0.234545  
cnt             1.000000

In [334]:
corr_m['cnt'].sort_values(ascending=False)

cnt               1.000000
atemp             0.631066
temp              0.627494
yr                0.566710
season            0.406100
mnth              0.279977
weekday           0.067443
workingday        0.061156
holiday          -0.068348
hum              -0.100659
windspeed(mph)   -0.234545
weathersit       -0.297391
Name: cnt, dtype: float64

Переменные atemp (0.63) и temp (0.62) сильнее всего коррелируют с целевой.

Между переменными atemp и temp существует сильная зависимость (0.991702). Чтобы устранить мультиколлинеарность удалим atemp (по моему мнению лучше оставить точную температуру, нежели по ощущениям).

In [336]:
df = df.drop(columns='atemp')

# Задание 10

Разделите данные на предикторы и целевую переменную. Обучите модель с помощью классической линейной регрессии. Линейную регрессию необходимо реализовать без готовых библиотек.

Оцените качество модели с помощью хотя бы двух метрик качества.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода. 

In [338]:
# разделение на предикторы и целевую переменную
X = np.array(df.drop(columns='cnt'))
y = np.array(df['cnt'])

# bias
X = np.concatenate([np.ones((X.shape[0], 1)), X], axis=1)

# коэффициенты линейной регрессии
beta = np.linalg.inv(X.T @ X) @ X.T @ y

# прогнозы
y_pred = X @ beta

# метрики
mse = np.mean((y - y_pred) ** 2)
r2 = 1 - (np.sum((y - y_pred) ** 2) / np.sum((y - np.mean(y)) ** 2))

print(f'Коэффициенты линейной регрессии: {beta}')
print(f'MSE: {mse}')
print(f'R^2: {r2}')

Коэффициенты линейной регрессии: [1626.11274728  515.23642917 2040.88251485  -39.80545835 -536.84058293
   67.21118086  119.21163986 -622.29155022  125.71269207   -9.60150621
  -40.75260927]
MSE: 753878.6752930074
R^2: 0.7988398641466311


In [339]:
df['cnt'].describe()

count     731.000000
mean     4504.348837
std      1937.211452
min        22.000000
25%      3152.000000
50%      4548.000000
75%      5956.000000
max      8714.000000
Name: cnt, dtype: float64

Среднее отклонение $\sqrt{MSE} \approx 868$ в сравнении со стандартным отклоненем указывает на хороший результат, но заметное отклонение остается.

$R^2$ указывает на то, что модель объясняет $\approx 80\%$ вариаций и достаточно хорошо описывает данные.